x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/job/process/autodetect/JobModelSnapshotUpgrader.java

```diff
@@ -358,30 +362,46 @@ private void checkResultsProcessorIsAlive() {
         }
 
         void shutdown(Exception e) {
+            logger.debug("[{}] [{}] shutdown initiated", jobId, snapshotId);
             // No point in sending an action to the executor if the process has died
             if (process.isProcessAlive() == false) {
+                logger.debug("[{}] [{}] process is dead, no need to shutdown", jobId, snapshotId);
                 onFinish.accept(e);
                 autodetectWorkerExecutor.shutdown();
                 stateStreamer.cancel();
                 return;
             }
-            autodetectWorkerExecutor.execute(() -> {
+            Future<?> future = autodetectWorkerExecutor.submit(() -> {
                 try {
+                    logger.debug("[{}] [{}] shutdown is now occurring", jobId, snapshotId);
                     if (process.isReady()) {
                         process.close();
                     } else {
                         processor.setProcessKilled();
                         process.kill(true);
-                        processor.awaitCompletion();
+                        stateStreamer.cancel();
                     }
+                    processor.awaitCompletion();
                 } catch (IOException | TimeoutException exc) {
                     logger.warn(() -> new ParameterizedMessage("[{}] [{}] failed to shutdown process", jobId, snapshotId), exc);
                 } finally {
                     onFinish.accept(e);
                 }
+                logger.debug("[{}] [{}] connection for upgrade has been closed, process is shutdown", jobId, snapshotId);
             });
-            autodetectWorkerExecutor.shutdown();
-            stateStreamer.cancel();
+            try {
+                future.get();
+                autodetectWorkerExecutor.shutdown();
```

Comment:
This is the key here. Previously, it COULD be that `shutdown` is processed in the window between looking for something in the queue and checking if its closed.

This ensures that the `shutdown` request task is handled before closing the executor